js/modules/k6/data/data.go

```diff
@@ -23,21 +23,50 @@ package data
 import (
 	"context"
 	"strconv"
+	"sync"
 
 	"github.com/dop251/goja"
 	"github.com/loadimpact/k6/js/common"
-	"github.com/loadimpact/k6/js/internal/modules"
 	"github.com/loadimpact/k6/lib"
 	"github.com/pkg/errors"
 )
 
-type data struct{}
+type data struct {
+	shared sharedArrays
+}
+
+type sharedArrays struct {
+	data map[string]sharedArray
+	mu   sync.RWMutex
+}
+
+func (s *sharedArrays) get(rt *goja.Runtime, name string, call goja.Callable) sharedArray {
+	s.mu.RLock()
+	array, ok := s.data[name]
+	s.mu.RUnlock()
+	if !ok {
+		s.mu.Lock()
+		array, ok = s.data[name]
+		if !ok {
```

Comment:
Why check for this twice? If it doesn't exist the first time couldn't you just lock and add it?

Also I'm not sure why you need the `func()` and `defer` below. Why not just call `Unlock()` after `s.data[name] = array`?

Reply:
First; I am not certain the RWLock is needed at all as I doubt it speeds up anything.

The second check is because the first check only happens with the read lock and it needs to be unlocked in order to get the write lock to add it. So it is possible for multiple VUs to get to that point and one to writelock, execute the callback, go to write it, unlock the write lock and then the second VU to come by and do the same all over again effectively executing the callback twice. Which is exactly what we *don't* want to happen.

The defer is needed as there could be a panic/goja exception in the callback code and we don't want to lock the whole `sharedArrays` forever just because of that.